### DISCUSSION TITLE
Clipper implementation

### DISCUSSION THREAD

**Discussion by BradleyChatha** [PlummersSoftwareLLC/Primes#668] [Q&A]
An acquaintance of mine would like to implement a solution in an ancient language known as Clipper.

Considering the language's age and nicheness, it's kind of difficult to find an implementation for it that will also run in docker.

I've made a **very** barebones implementation for it at https://github.com/BradleyChatha/clipper-lite which transpiles the language into D, and I was just wondering a few things after confirming it is able to produce the sieve.

* Would a transpiled implementation count as 'legal' or faithful still? I'm asking since I'm fairly certain this acquaintance of mine would love to have the solution on the leaderboard.
* Should I embed the transpiler (3 F# files) into the solution folder and run it from there in docker?
* While there does seem to be semi-modern implementations via the Harbour and xHarbour projects, I'm very uncertain about their ability to be used within a docker container, and just in general I'm quite uncertain about them. Is it a hard requirement that I should use these pre-existing projects rather than this transpiler?
* Are my modifications to the language (listed in the repo's README) deal breakers regarding faithfulness?

Thank you for your time and answers.

**Comment by rbergen** [comment]
I had a look at Harbour, and I got it installed and running within a Docker container with the following Dockerfile:
```dockerfile
FROM ubuntu:20.04

ENV HARBOUR_PKG=r2017-12-15-18_53_harbour_3.2.0-1_amd64.deb

WORKDIR /opt

RUN apt-get update && \
    apt-get install -y wget && \
    wget "https://sourceforge.net/projects/harbour-project/files/binaries-linux-ubuntu/nightly/${HARBOUR_PKG}/download" -O ${HARBOUR_PKG} && \
    apt-get install -y "./${HARBOUR_PKG}" && \
    rm "${HARBOUR_PKG}"
```

I guess what I'm saying with that is that it does not seem to be particularly difficult to get Harbour running within Docker, at least. An important side note to make is that I have not yet actually compiled a Clipper program with it.
You also say that you're uncertain about it and xHarbour in general, but I can't deduce from your comment why that is.

I would say that in principle, contributors have reasonable freedom towards selecting the compiler for their solutions, whether that is a transpiler or not. We already have a number of languages/solutions in the repo that use a transpiler within their build process, although generally speaking that tends to happen under the hood. 

However, with at least one recognized open source compiler available for Clipper, provided it actually compiles, I do think we can and should take "standard Clipper" as the frame of reference. To make that concrete: I would be fine with accepting a Clipper solution that is actually built with a hand-built transpiler that is also part of the solution, as long as it can also be compiled with a functioning installation of (x)Harbour.

**Reply by BradleyChatha** [reply]
I'll fiddle around with trying to get something compiled and running in docker.

As for my distrust for things like Harbour, I don't know, I tend to feel wary about projects for old, niche things, especially when the projects themselves have seemed to stop updating (read as: a PITA to compile even for things that are only a year or so 'out of date').

It's even worse when the person who wants to use Clipper keeps going on about whipping out a computer older than me to run a compiler older than me to do it in ;^) I'm not 100% sure he's even aware of the docker requirement, which is also why I was tempted into making a transpiler rather than messing around trying to get other people's seemingly abandoned projects (Harbour) to work.

I'll see what I can do with that docker image though, and if I'm not successful then I'll fall back to the transpiler. Thank you for looking into it as well!

**Reply by BradleyChatha** [reply]
We do appear to have a possible issue here:

![Screenshot from 2021-08-21 18-35-33](https://user-images.githubusercontent.com/3835574/130330373-b935e3b7-cc1c-4194-9ac7-47bbe3daaa3b.png)

Because Clipper is kind of like "ncurses - the language" it clears the screen with that gray background colour (assuming it uses ANSI sequences) and *then* it prints out the Hello world text.

I imagine that the tool for parsing the program output would choke on this?

**Reply by marghidanu** [reply]
It shouldn’t, try it out.

**Reply by BradleyChatha** [reply]
```
localhost:/home/brad/Desktop/Primes # make DIRECTORY=PrimeClipper/solution_1/
info: Unconfined mode: false
info: Detected architecture: amd64
info: [PrimeClipper][solution_1] Building...
error: [PrimeClipper][solution_1] Failed building solution!
error: No data was collected!
```

**Reply by BradleyChatha** [reply]
I *assume* the "failed building solution" is **because** it couldn't collect data, rather than it actually failling the build the docker image.

For reference I can build and run the docker image perfectly fine.

Also for reference: The string I'm writing out is "BradleyChatha-Single-SieveCT;7780;5.00036;1;algorithm=base,bits=1,faithful=no" which I took from the D solution, so it should be in correct form.

**Reply by marghidanu** [reply]
No, it actually failed building the Docker image. You’re not even running the solution. I’m curious how you are building the Docker image. Please put in a PR, it makes no sense to use a GitHub discussion for this.
